{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import math\n",
    "import numpy as np\n",
    "np.random.seed(0)\n",
    "import sys\n",
    "import os\n",
    "import scipy.sparse\n",
    "np.set_printoptions(threshold=sys.maxsize)\n",
    "\n",
    "current = os.path.dirname(os.path.realpath('__file__'))\n",
    "parent = os.path.dirname(current)\n",
    "parent = os.path.dirname(parent)\n",
    "parent = os.path.dirname(parent)\n",
    "\n",
    "\n",
    "sys.path.append(parent)\n",
    "\n",
    "import benchmark\n",
    "import tsp_qubo\n",
    "import plotter\n",
    "import annealing\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Parameters():\n",
    "    '''Container for Qhea parameters'''\n",
    "    def __init__(self, optimiziation_rate, n_individuals, n_generations, selection_pressure, bias):\n",
    "        self.optimization_rate = optimiziation_rate\n",
    "        self.n_individuals = n_individuals\n",
    "        self.n_generations = n_generations\n",
    "        self.selection_pressure = selection_pressure\n",
    "        self.bias = bias"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Data():\n",
    "    '''Container to collect data from the algorithm'''\n",
    "    fitness = 0\n",
    "    value = 1\n",
    "    def __init__(self, individual):\n",
    "        self.evolution = np.array([individual], dtype=object)\n",
    "        self.all_time_fittest = individual.copy()\n",
    "\n",
    "    def add_individual(self, individual):\n",
    "        self.evolution = np.vstack([self.evolution, individual])\n",
    "        if self.all_time_fittest[self.fitness] > individual[self.fitness]:\n",
    "            self.all_time_fittest = individual.copy()\n",
    "\n",
    "    def get_evolution(self):\n",
    "        return self.evolution\n",
    "    \n",
    "    def get_fitness_evolution(self):\n",
    "        return self.evolution[:, self.fitness]\n",
    "    \n",
    "    def get_individual_evolution(self):\n",
    "        return self.evolution[:, self.value]\n",
    "    \n",
    "    def get_solution(self):\n",
    "        return self.all_time_fittest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Qhea():\n",
    "    fitness = 0 # variable used to index individual for fitness\n",
    "    value = 1 # variable used to index individual for value, eg. chromosome\n",
    "    def __init__(self, qubo, parameters):\n",
    "        '''\n",
    "        :param qubo: qubo matrix  to optimize (minimize)\n",
    "        :param parameters: Parameter oject with params to execute the algorithm\n",
    "        '''\n",
    "        self.qubo = qubo # save qubo\n",
    "        self.qubo_T = qubo.T\n",
    "        self.sparse = scipy.sparse.coo_array(qubo)\n",
    "        self.n_vars = len(qubo) # n_vars is the chromosome size, 2 ** n_vars is the solution space\n",
    "        self.n_nodes = int(math.sqrt(self.n_vars))\n",
    "        self.n_individuals = parameters.n_individuals # size of the population\n",
    "        self.n_generations = parameters.n_generations # number of generations \n",
    "        self.selection_pressure = parameters.selection_pressure\n",
    "        self.bias = parameters.bias\n",
    "        self.optimization_rate = parameters.optimization_rate # probability of local_optimiztaion\n",
    "        self.population, self.crossover_sequence, self.data, self.probabilities = self.init() # init data\n",
    "        \n",
    "        \n",
    "    def init(self):\n",
    "        population = np.array([[0, np.random.randint(0, 2, self.n_vars)] for j in range(self.n_individuals)],\n",
    "                      dtype=object) # init population\n",
    "        \n",
    "        for ind in range(self.n_individuals):\n",
    "            population[ind][self.fitness] = self.cost_function(population[ind][self.value]) # initial fitness to population\n",
    "        \n",
    "        population = population[population[:, self.fitness].argsort()] # sort ascending by fitness\n",
    "        # self.clusters = self.get_sequences() # in this experiment, clusters serve as local optimization genomes\n",
    "        crossover_sequence = np.array([0^(int(i / int(self.n_vars / 2)) > 0) for i in range(self.n_vars)])\n",
    "        data = Data(population[0]) # init data object to collect data from algorithm\n",
    "        \n",
    "        probabilities = []\n",
    "\n",
    "        for i in range(1, self.n_individuals+1):\n",
    "            p = (1 / self.n_individuals) * (self.selection_pressure - (2 * self.selection_pressure - 2) * (\n",
    "                (i - 1) / (self.n_individuals - 1)))\n",
    "            probabilities.append(p)\n",
    "        return population, crossover_sequence, data, probabilities\n",
    "    \n",
    "    def cost_function(self, val):\n",
    "        return val @ self.sparse @ val # calculate fitness\n",
    "\n",
    "\n",
    "    def derive_sub_qubo(self,slicing_equence, chromosome):\n",
    "        sub_qubo = self.slice_qubo(slicing_equence,slicing_equence)\n",
    "        freezing_sequence = np.logical_not(slicing_equence)\n",
    "        freezed_genes = np.logical_and(freezing_sequence, chromosome)\n",
    "        if np.sum(freezed_genes) != 0:\n",
    "\n",
    "            sub_qubo_2 = np.transpose(self.slice_qubo(slicing_equence, freezed_genes))\n",
    "            sub_qubo_3 = self.slice_qubo(freezed_genes, slicing_equence)\n",
    "            \n",
    "            sum = sub_qubo_2 + sub_qubo_3\n",
    "            sum_axis = np.sum(sum, 0)\n",
    "            diag = np.diag(sum_axis)\n",
    "            sub_qubo += diag\n",
    "\n",
    "        return sub_qubo\n",
    "    \n",
    "\n",
    "    def local_optimization(self, chromosome):\n",
    "        r = np.random.uniform(0,1)\n",
    "        if r < self.optimization_rate:            \n",
    "            for i in range(len(chromosome)):\n",
    "                row = self.qubo[i]\n",
    "                column = self.qubo[:, i]\n",
    "                \n",
    "                chromosome[i] = 1\n",
    "                sums = np.add(row,column)\n",
    "                if np.dot(chromosome,sums) - self.qubo[i][i] > 0:\n",
    "                    chromosome[i] = 0\n",
    "                \n",
    "        return chromosome\n",
    "\n",
    "  \n",
    "\n",
    "\n",
    "    def optimize_sub_problem(self, sequence, chromosome):\n",
    "        sub_qubo = self.derive_sub_qubo(sequence, chromosome)\n",
    "        solution = annealing.d_wave.anneal(sub_qubo)[0].astype(int)\n",
    "        s_i = np.where(sequence == 1)[0]\n",
    "        result = np.array([chromosome[i] if i not in s_i else solution[np.where(s_i == i)[0][0]] for i in range(len(chromosome))])\n",
    "        return result\n",
    "    \n",
    "    def slice_qubo(self,x,y):\n",
    "        return self.qubo[np.ix_( np.where(x==1)[0], np.where(y==1)[0])]\n",
    "    \n",
    "\n",
    "    def uniform_crossover_sequence(self):\n",
    "        probabilities = np.random.uniform(0.0, 1.0, self.n_vars)\n",
    "\n",
    "        return probabilities > self.bias\n",
    "\n",
    "    def optimize(self):\n",
    "\n",
    "        for gen in range(self.n_generations): # evolution loop\n",
    "            self.mutation_rate = 1 / (gen + 1)\n",
    "            for ind in range(int(self.n_individuals / 2)): # iterating over the better half of individuals\n",
    "                \n",
    "                # mutation sequence is an array of [0,1], with a probabilty of mutation_rate for each index to be 1\n",
    "                mutation_sequence_1 = np.random.random(self.n_vars) < self.mutation_rate # mutation sequence 1\n",
    "                mutation_sequence_2 = np.random.random(self.n_vars) < self.mutation_rate # mutation sequence 2\n",
    "\n",
    "                                 # selecting parents probabilistic\n",
    "                parents = np.array([[0, np.random.randint(0, 2, self.n_vars)] for j in range(2)],\n",
    "                      dtype=object)\n",
    "                \n",
    "                r = np.random.uniform(0,1,1)\n",
    "\n",
    "                index_1 = (np.cumsum(self.probabilities)>r).argmax()\n",
    "                parents[0] = self.population[index_1]\n",
    "\n",
    "                r = np.random.uniform(0,1,1)\n",
    "         \n",
    "                # select parent_2        \n",
    "                index_2 = (np.cumsum(self.probabilities)>r).argmax()\n",
    "                parents[1] = self.population[index_2]\n",
    "                # mutation sequence is an array of [0,1], with a probabilty of mutation_rate for each index to be 1\n",
    "                mutation_sequence_1 = np.random.random(self.n_vars) < self.mutation_rate # mutation sequence 1\n",
    "                mutation_sequence_2 = np.random.random(self.n_vars) < self.mutation_rate # mutation sequence 2\n",
    "                self.crossover_sequence = self.uniform_crossover_sequence()\n",
    "                # crossover with selected parents (first child), then mutation\n",
    "                offspring_1_chromosome = np.logical_xor(\n",
    "                np.add(\n",
    "                    np.logical_and(parents[0][self.value], self.crossover_sequence), # getting the genes from first parent (where crossover is 1)\n",
    "                    np.logical_and(parents[1][self.value], np.logical_xor(self.crossover_sequence, np.ones(self.n_vars))) # getting the genes from second parent (where crossover is 0)\n",
    "                ),mutation_sequence_1)\n",
    "\n",
    "                self.crossover_sequence = self.uniform_crossover_sequence()\n",
    "                # crossover with selected parents (second child), then mutation\n",
    "                offspring_2_chromosome = np.logical_xor(\n",
    "                np.add(\n",
    "                    np.logical_and(parents[1][self.value], self.crossover_sequence), # getting the genes from second parent (where crossover is 1)\n",
    "                    np.logical_and(parents[0][self.value], np.logical_xor(self.crossover_sequence, np.ones(self.n_vars)))  # getting the genes from first parent (where crossover is 0)\n",
    "                ), mutation_sequence_2)\n",
    "                # note how the order of parents changed for individual 2\n",
    "                self.population = np.vstack([self.population, np.array([0, offspring_1_chromosome], dtype=object)], dtype=object)\n",
    "                self.population = np.vstack([self.population, np.array([0, offspring_2_chromosome], dtype=object)], dtype=object)\n",
    "\n",
    "                offspring_1_chromosome = self.local_optimization(offspring_1_chromosome)\n",
    "                offspring_2_chromosome = self.local_optimization(offspring_2_chromosome)\n",
    "                \n",
    "                \n",
    "            # apply fitness to every new individual\n",
    "            for ind in range(self.n_individuals, len(self.population)):\n",
    "                self.population[ind][self.fitness] = self.cost_function(self.population[ind][self.value])\n",
    "            \n",
    "            # sort ascending by fitness\n",
    "            self.population = self.population[self.population[:, self.fitness].argsort()]\n",
    "            self.population = self.population[:self.n_individuals]\n",
    "            # add fittest individual to data set\n",
    "            self.data.add_individual(self.population[0])\n",
    "        \n",
    "        # when done, return data set\n",
    "        return self.data\n",
    "    \n",
    "    def get_clusters(self):\n",
    "        clusters = np.zeros((self.n_nodes, self.n_nodes))\n",
    "        cluster_c = int(self.n_nodes / self.cluster_size)\n",
    "\n",
    "        for n in range(self.n_clusters):\n",
    "            amp = np.zeros((cluster_c, cluster_c))\n",
    "            amp[n % cluster_c][int(n / cluster_c)] = 1\n",
    "            cluster = np.ones((self.cluster_size, self.cluster_size)) * n\n",
    "            clusters += np.kron(amp, cluster)\n",
    "\n",
    "        return np.reshape(clusters, self.n_vars)\n",
    "\n",
    "\n",
    "    def get_sequences(self):\n",
    "        \n",
    "        clusters = self.get_clusters()\n",
    "        cluster_sequences = np.array([np.zeros(self.n_vars) for i in range(self.n_clusters)])\n",
    "        for i in range(self.n_clusters):\n",
    "            cluster_sequences[i] = np.array([0 ^ (i == int(clusters[j])) for j in range(self.n_vars)])\n",
    "        \n",
    "        \n",
    "        return cluster_sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "tsp = benchmark.distances.get_tsp(benchmark.distances.get_tsp_names()[4]).distances\n",
    "n_nodes = len(tsp)\n",
    "penalty = 100\n",
    "constraint_matrix = np.zeros((n_nodes ** 2, n_nodes ** 2))\n",
    "constraint_matrix = tsp_qubo.generate.apply_penalties(constraint_matrix, penalty)\n",
    "\n",
    "amp = np.zeros((n_nodes, n_nodes))\n",
    "for i in range(len(amp)):\n",
    "    amp[i][(i + 1) % n_nodes] = 1\n",
    "objective = np.kron(amp, tsp)\n",
    "qubo = objective + constraint_matrix\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[1;32mc:\\Users\\brend\\Desktop\\brendans-masterarbeit\\experiment_archiv\\tsp_50_nodes_2500_vars\\local_optimization\\one_iterative_ordered_0.5.ipynb Cell 6\u001b[0m in \u001b[0;36m<cell line: 11>\u001b[1;34m()\u001b[0m\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/brend/Desktop/brendans-masterarbeit/experiment_archiv/tsp_50_nodes_2500_vars/local_optimization/one_iterative_ordered_0.5.ipynb#W5sZmlsZQ%3D%3D?line=10'>11</a>\u001b[0m \u001b[39mfor\u001b[39;00m i \u001b[39min\u001b[39;00m \u001b[39mrange\u001b[39m(\u001b[39m5\u001b[39m):\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/brend/Desktop/brendans-masterarbeit/experiment_archiv/tsp_50_nodes_2500_vars/local_optimization/one_iterative_ordered_0.5.ipynb#W5sZmlsZQ%3D%3D?line=11'>12</a>\u001b[0m     solver \u001b[39m=\u001b[39m Qhea(qubo\u001b[39m=\u001b[39mqubo,parameters\u001b[39m=\u001b[39mparameters)\n\u001b[1;32m---> <a href='vscode-notebook-cell:/c%3A/Users/brend/Desktop/brendans-masterarbeit/experiment_archiv/tsp_50_nodes_2500_vars/local_optimization/one_iterative_ordered_0.5.ipynb#W5sZmlsZQ%3D%3D?line=12'>13</a>\u001b[0m     result \u001b[39m=\u001b[39m solver\u001b[39m.\u001b[39;49moptimize()\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/brend/Desktop/brendans-masterarbeit/experiment_archiv/tsp_50_nodes_2500_vars/local_optimization/one_iterative_ordered_0.5.ipynb#W5sZmlsZQ%3D%3D?line=13'>14</a>\u001b[0m     evolution \u001b[39m=\u001b[39m result\u001b[39m.\u001b[39mget_fitness_evolution()\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/brend/Desktop/brendans-masterarbeit/experiment_archiv/tsp_50_nodes_2500_vars/local_optimization/one_iterative_ordered_0.5.ipynb#W5sZmlsZQ%3D%3D?line=14'>15</a>\u001b[0m     evolutions \u001b[39m=\u001b[39m np\u001b[39m.\u001b[39madd(evolutions, evolution)\n",
      "\u001b[1;32mc:\\Users\\brend\\Desktop\\brendans-masterarbeit\\experiment_archiv\\tsp_50_nodes_2500_vars\\local_optimization\\one_iterative_ordered_0.5.ipynb Cell 6\u001b[0m in \u001b[0;36mQhea.optimize\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    <a href='vscode-notebook-cell:/c%3A/Users/brend/Desktop/brendans-masterarbeit/experiment_archiv/tsp_50_nodes_2500_vars/local_optimization/one_iterative_ordered_0.5.ipynb#W5sZmlsZQ%3D%3D?line=139'>140</a>\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mpopulation \u001b[39m=\u001b[39m np\u001b[39m.\u001b[39mvstack([\u001b[39mself\u001b[39m\u001b[39m.\u001b[39mpopulation, np\u001b[39m.\u001b[39marray([\u001b[39m0\u001b[39m, offspring_2_chromosome], dtype\u001b[39m=\u001b[39m\u001b[39mobject\u001b[39m)], dtype\u001b[39m=\u001b[39m\u001b[39mobject\u001b[39m)\n\u001b[0;32m    <a href='vscode-notebook-cell:/c%3A/Users/brend/Desktop/brendans-masterarbeit/experiment_archiv/tsp_50_nodes_2500_vars/local_optimization/one_iterative_ordered_0.5.ipynb#W5sZmlsZQ%3D%3D?line=141'>142</a>\u001b[0m     offspring_1_chromosome \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mlocal_optimization(offspring_1_chromosome)\n\u001b[1;32m--> <a href='vscode-notebook-cell:/c%3A/Users/brend/Desktop/brendans-masterarbeit/experiment_archiv/tsp_50_nodes_2500_vars/local_optimization/one_iterative_ordered_0.5.ipynb#W5sZmlsZQ%3D%3D?line=142'>143</a>\u001b[0m     offspring_2_chromosome \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mlocal_optimization(offspring_2_chromosome)\n\u001b[0;32m    <a href='vscode-notebook-cell:/c%3A/Users/brend/Desktop/brendans-masterarbeit/experiment_archiv/tsp_50_nodes_2500_vars/local_optimization/one_iterative_ordered_0.5.ipynb#W5sZmlsZQ%3D%3D?line=145'>146</a>\u001b[0m \u001b[39m# apply fitness to every new individual\u001b[39;00m\n\u001b[0;32m    <a href='vscode-notebook-cell:/c%3A/Users/brend/Desktop/brendans-masterarbeit/experiment_archiv/tsp_50_nodes_2500_vars/local_optimization/one_iterative_ordered_0.5.ipynb#W5sZmlsZQ%3D%3D?line=146'>147</a>\u001b[0m \u001b[39mfor\u001b[39;00m ind \u001b[39min\u001b[39;00m \u001b[39mrange\u001b[39m(\u001b[39mself\u001b[39m\u001b[39m.\u001b[39mn_individuals, \u001b[39mlen\u001b[39m(\u001b[39mself\u001b[39m\u001b[39m.\u001b[39mpopulation)):\n",
      "\u001b[1;32mc:\\Users\\brend\\Desktop\\brendans-masterarbeit\\experiment_archiv\\tsp_50_nodes_2500_vars\\local_optimization\\one_iterative_ordered_0.5.ipynb Cell 6\u001b[0m in \u001b[0;36mQhea.local_optimization\u001b[1;34m(self, chromosome)\u001b[0m\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/brend/Desktop/brendans-masterarbeit/experiment_archiv/tsp_50_nodes_2500_vars/local_optimization/one_iterative_ordered_0.5.ipynb#W5sZmlsZQ%3D%3D?line=67'>68</a>\u001b[0m column \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mqubo[:, i]\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/brend/Desktop/brendans-masterarbeit/experiment_archiv/tsp_50_nodes_2500_vars/local_optimization/one_iterative_ordered_0.5.ipynb#W5sZmlsZQ%3D%3D?line=69'>70</a>\u001b[0m chromosome[i] \u001b[39m=\u001b[39m \u001b[39m1\u001b[39m\n\u001b[1;32m---> <a href='vscode-notebook-cell:/c%3A/Users/brend/Desktop/brendans-masterarbeit/experiment_archiv/tsp_50_nodes_2500_vars/local_optimization/one_iterative_ordered_0.5.ipynb#W5sZmlsZQ%3D%3D?line=70'>71</a>\u001b[0m sums \u001b[39m=\u001b[39m np\u001b[39m.\u001b[39;49madd(row,column)\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/brend/Desktop/brendans-masterarbeit/experiment_archiv/tsp_50_nodes_2500_vars/local_optimization/one_iterative_ordered_0.5.ipynb#W5sZmlsZQ%3D%3D?line=71'>72</a>\u001b[0m \u001b[39mif\u001b[39;00m np\u001b[39m.\u001b[39mdot(chromosome,sums) \u001b[39m-\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mqubo[i][i] \u001b[39m>\u001b[39m \u001b[39m0\u001b[39m:\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/brend/Desktop/brendans-masterarbeit/experiment_archiv/tsp_50_nodes_2500_vars/local_optimization/one_iterative_ordered_0.5.ipynb#W5sZmlsZQ%3D%3D?line=72'>73</a>\u001b[0m     chromosome[i] \u001b[39m=\u001b[39m \u001b[39m0\u001b[39m\n",
      "\u001b[1;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "parameters = Parameters(\n",
    "    optimiziation_rate=0.005,\n",
    "    n_individuals=100,\n",
    "    n_generations=1000,\n",
    "    selection_pressure=2,\n",
    "    bias=0.5\n",
    ")\n",
    "\n",
    "evolutions = np.array([0 for i in range(parameters.n_generations+1)], dtype=object)\n",
    "histogramm  = []\n",
    "for i in range(5):\n",
    "    solver = Qhea(qubo=qubo,parameters=parameters)\n",
    "    result = solver.optimize()\n",
    "    evolution = result.get_fitness_evolution()\n",
    "    evolutions = np.add(evolutions, evolution)\n",
    "    histogramm.append(result.all_time_fittest[0])\n",
    "\n",
    "evolutions = evolutions / 5\n",
    "print(repr(histogramm))\n",
    "plotter.plot_evolution(evolutions, \"one_iterative_ordered_0.5%\", 2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Experiment Description\n",
    "The Algorithm was tested with a medium mutation rate of 5% probability. \n",
    "\n",
    "## Result\n",
    "<img src=\"files/local_optimization_cluster_large.png\">"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  },
  "vscode": {
   "interpreter": {
    "hash": "ccfc4daa9a338a32d204ce04182463af6af428de08f13a8e09320d9ab6a889ea"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
